### DISCUSSION TITLE
Enabling ML by default

### DISCUSSION THREAD

**Discussion by shyamvalsan** [netdata/netdata#12130] [Ideas]
Hey @andrewm4894 @vkalintiris @ktsaou @cpipilas 

Creating this discussion for us to get to the immediate actions required to enable ML by default. Please share your feedback (and add others you think should be on this discussion thread). 

Let's try to arrive at the list of actions by 16-Feb-2022

### **Objective**

- Enable ML by default. This implies enabling anomaly detection by default for all possible metrics, metric correlation continues to be on-click-ML triggered by user.
- Ensure that resource and performance utilization remain within acceptable limits when ML is enabled on the edge.
- Choose the right default configuration parameters for ML to safely allow it to run everywhere all the time.

**Comment by ktsaou** [comment]
> * Enable ML by default. This implies enabling anomaly detection by default for all possible metrics, metric correlation continues to be on-click-ML triggered by user

Metric correlation is not related at all to ML. Currently it is also centralized, so as far as the agents are concerned, the current metric correlation is concerned, it is just `/api/v1/data` queries. Right?

> * Ensure that resource and performance utilization remain within acceptable limits when ML is enabled on the edge

We need to test this. Any ideas what tests we should run?

> * Choose the right default configuration parameters for ML to safely allow it to run everywhere all the time

What are the current settings?

**Reply by shyamvalsan** [reply]
The current [ML config settings](https://learn.netdata.cloud/guides/monitor/anomaly-detection) are as below

```
[ml]
    # enabled = no
    # maximum num samples to train = 14400
    # minimum num samples to train = 3600
    # train every = 3600
    # num samples to diff = 1
    # num samples to smooth = 3
    # num samples to lag = 5
    # maximum number of k-means iterations = 1000
    # dimension anomaly score threshold = 0.99
    # host anomaly rate threshold = 0.01000
    # minimum window size = 30.00000
    # maximum window size = 600.00000
    # idle window size = 30.00000
    # window minimum anomaly rate = 0.25000
    # anomaly event min dimension rate threshold = 0.05000
    # hosts to skip from training = !*
    # charts to skip from training = !system.* !cpu.* !mem.* !disk.* !disk_* !ip.* !ipv4.* !ipv6.* !net.* !net_* !netfilter.* !services.* !apps.* !groups.* !user.* !ebpf.* !netdata.* *
```

**Reply by ktsaou** [reply]
In simple words, we train every hour (`train every = 3600`), 4 hours of data in the past (`maximum num samples to train = 14400`).
Is there anything else that influences cpu or memory consumption?

**Reply by shyamvalsan** [reply]
Regarding Metric Correlation - yes it is not linked to ML config at all as far as I know (@andrewm4894 to confirm) but my question (or concern) was more about conflating ML with Anomaly Detection in terms of config going forward. 

For now, this maybe true and for the purpose of this discussion maybe that's good enough.

**Reply by vkalintiris** [reply]
> Is there anything else that influences cpu or memory consumption?

Yes, `charts to skip from training`. The more charts we skip from training the lighter ML will be on CPU/RAM/disk/etc.

Also, the following list of configuration options affect performance:

```
    # num samples to diff = 1
    # num samples to smooth = 3
    # num samples to lag = 5
    # maximum number of k-means iterations = 1000
```

However, we don't expect users to modify these values unless they are ML experts.

**Reply by ktsaou** [reply]
> Regarding Metric Correlation - yes it is not linked to ML config at all as far as I know (@andrewm4894 to confirm) but my question (or concern) was more about conflating ML with Anomaly Detection in terms of config going forward.

ML and Anomaly Detection is not related to metric correlations. I still don't get this.

**Reply by shyamvalsan** [reply]
@ktsaou What metrics are being analyzed for anomaly detection would be one major item that influences cpu or memory. Do we want all 1500-2000 dimensions to be analyzed by default - or a subset, especially considering lower resource agents.  

I think some of the other configs might affect resources as well, but to a lower extent that `train every` and `maximum num samples to train`

@vkalintiris @andrewm4894 thoughts?

**Reply by shyamvalsan** [reply]
> > Regarding Metric Correlation - yes it is not linked to ML config at all as far as I know (@andrewm4894 to confirm) but my question (or concern) was more about conflating ML with Anomaly Detection in terms of config going forward.
> 
> ML and Anomaly Detection is not related to metric correlations. I still don't get this.

Let me try and clear up the confusion. My point was if we end up using ML for other things in the future should it all live under this parent config we are turning ON by default now. Or should this config be specific to "anomaly detection".

**Reply by ktsaou** [reply]
Great!

About the metrics, I think we should exclude all metrics that are not useful for anomaly detection. These could be:

1. metrics that are noisy by nature and may not provide any useful info to the user. These could be the hardware interrupts, the netdata section, etc.
2. metrics that count error conditions, like errors on disks, errors on network interfaces, etc. These metrics should have traditional alarms attached to them to raise when we have errors.
3. metrics that are mainly idle, like a network interface that does not have any traffic, a disk that is just sitting idle, etc. For example, to save cpu resources, we could do an `stddev` over the last 4 hours on all metrics, and only when there is some action in it to run anomaly detection. This will probably exclude more than 50% of the metrics. Even better, we could calculate the variability of metrics during data collection, so that the agent has a list of candidate metrics right away.

> ```
>     # charts to skip from training = !system.* !cpu.* !mem.* !disk.* !disk_* !ip.* !ipv4.* !ipv6.* !net.* !net_* !netfilter.* !services.* !apps.* !groups.* !user.* !ebpf.* !netdata.* *
> ```

I am not sure why we exclude `system.*` or `apps.*` or `disk.*` or `net.*`. We shouldn't detect anomalies on system resources utlization?

**Reply by ktsaou** [reply]
Updated the comment above many times. Read it again please.

**Reply by vkalintiris** [reply]
> I am not sure why we exclude system.* or apps.* or disk.* or net.*

We don't exclude them. The simple pattern follows this format `charts to skip from training = !chart_a !chart_b *`, ie. do not train any charts except for `chart_a` and `chart_b`.

**Reply by ktsaou** [reply]
Oh! Yes, wrong interpretation from my side.

**Reply by shyamvalsan** [reply]
> Even better, we could calculate the variability of metrics during data collection, so that the agent has a list of candidate metrics right away.

I like this - an initial learning period during data collection to do this. And then maybe do it periodically (not every hour or every 4 hours, but maybe once a day check if the initial assumptions about those metrics still hold true).  This should trim the list of candidate metrics by > 50%

Do we want the following to be included? Does it provide a lot of user value? 

`!services.* !apps.* !groups.* !user.* !ebpf.* !netdata.*`

**Reply by ktsaou** [reply]
Then, we don't train any application specific metrics. Where is mysql, nginx, web_log?

**Reply by shyamvalsan** [reply]
Oh, yes. I need to go understand better about what raw metrics actually fall into these categories.

**Reply by vkalintiris** [reply]
> I like this - an initial learning period during data collection to do this. And then maybe do it periodically (not every hour or every 4 hours, but maybe once a day check if the initial assumptions about those metrics still hold true). This should trim the list of candidate metrics by > 50%

In theory, the only reason we are discussing this topic is because we use an offline clustering algorithm. Finding/implementing/testing custom heuristics that try to trim down the list of dimensions we want to train, is a time consuming and complex process that ultimately we wouldn't have to go through if we were to switch to an online algorithm.

**Reply by ktsaou** [reply]
> Let me try and clear up the confusion. My point was if we end up using ML for other things in the future should it all live under this parent config we are turning ON by default now. Or should this config be specific to "anomaly detection".

I can't answer that. It depends on what we are going to add. I think metric correlations will not have any configuration. All its parameters should be supplied with the query.

**Reply by ktsaou** [reply]
> In theory, the only reason we are discussing this topic is because we use an offline clustering algorithm. Finding/implementing/testing custom heuristics that try to trim down the list of dimensions we want to train, is a time consuming and complex process that ultimately we wouldn't have to go through if we were to switch to an online algorithm.

@vkalintiris let's start with something simple. If a metric is totally constant (i.e. `min = max` for the whole duration), why to train a model for it? How many metrics can we exclude with this? There will be many metrics that are just constant... Can we find out how many?

**Reply by shyamvalsan** [reply]
@vkalintiris I think we need to do a POC with an online approach to answer that for sure, maybe using something like a [Mini Batch K means](https://scikit-learn.org/stable/modules/generated/sklearn.cluster.MiniBatchKMeans.html) - another option would be to measure approximate entropy for each metric and have a threshold within which we think Anomaly Detection is valuable.

**Reply by andrewm4894** [reply]
> Is there anything else that influences cpu or memory consumption?

`charts to skip from training` is the main one - this is why the defaults here is like an allow list of typical default charts. Users would then need to add to it if wanted to expand to more charts.

**Reply by ktsaou** [reply]
> `charts to skip from training` is the main one - this is why the defaults here is like an allow list of typical default charts. Users would then need to add to it if wanted to expand to more charts.

hm... ideally it should be automatic.
This is why I suggested to eliminate all the ones that are not useful for ml.

**Reply by vkalintiris** [reply]
> How many metrics can we exclude with this? There will be many metrics that are just constant... Can we find out how many?

I built an agent and let it run overnight on an idle machine. I used the `/data` endpoint on every chart to figure out the number of constant dimensions. They are in the range of 40%-50%.

> If a metric is totally constant (i.e. min = max for the whole duration), why to train a model for it?

1. This optimization does not apply to the 1st training iteration. We'd have to build a model for the constant dimensions, because subsequent values might differ. As long as the values remain constant, we can skip retraining.
2. Somewhat obvious but for completeness sake - the peak CPU usage is not reduced. Only the total CPU consumption over the training window.

My intention is to:

1. Confirm the results we got from the `/data` endpoint, by checking directly the internal storage numbers of each dimension.
2. Implement the optimization, compare it against the current implementation, and test it on staging.

First, I'll prioritize addressing review comments in https://github.com/netdata/netdata/pull/12083, in order to unblock Andrew and the frontend team.

I'll share any updates here.

**Reply by shyamvalsan** [reply]
Thanks @vkalintiris this sounds like a good plan of action. 

And about this

> This optimization does not apply to the 1st training iteration. We'd have to build a model for the constant dimensions, because subsequent values might differ. As long as the values remain constant, we can skip retraining.

just thinking out aloud, can we not have a 0th iteration where we do not train the model but just observe and exclude constant dimensions? And then training starts in the next iteration?

**Comment by andrewm4894** [comment]
For me the following steps would need to happen before we consider enabling by default.

1. anomaly rate on any chart implemented on frontend.
2. one month of full active, internal, company wide, dogfooding.
3. after internal dogfooding, one month of usage and feedback by early adopters in the netdata community who have enabled the feature.
4. some rules around conditions that have to be met by the node for this feature to be activated, otherwise it should stay default off.

**Reply by ktsaou** [reply]
@andrewm4894 what do you expect to get out of these? What are you concerned about and how these steps will help?

I am interested about 2 and 3.

I think 1 will be implemented in the mean-time.
And 4 is accurate. We have to figure out these conditions.

But as I understand it, the first and key step to proceed further, should be understanding the CPU impact this has. Without it, we are just theorizing...

**Reply by andrewm4894** [reply]
1,2,3 are all just about getting as much exposure as possible to get as much qualitative feedback on the actual ml in a wide range of nodes and settings. From that we build our knowledge of where it works well, and where its maybe weaker and we need to improve things under the hood a little more. Then we would have a more informed viewpoint to decide if its "ready" or "good enough" for enabling by default as the first generation of the models.

**Reply by ktsaou** [reply]
I think that if the cpu impact is not significant, we should release it enabled by default as soon as possible. It is better to interact with real users instead of keeping it for us for so long. The number of use cases that will be revealed from users is enormous compared to the ones we will find.

So, the key question is what is the impact. We should do the most to minimize the impact. This is why intelligently selecting the metrics to be trained is very important.

**Comment by shyamvalsan** [comment]
To sum up the discussion.

This feature is planned for early Q2 along with ["anomaly rate on every chart" ](https://github.com/netdata/product/issues/2788) this  gives us a month or more of internal dogfooding and hopefully a few weeks of early adopter testing as well. 

In the mean time we can answer the question about the impact and how to minimize it.

**Reply by ktsaou** [reply]
@shyamvalsan @cpipilas please open a ticket about measuring the impact. We need hard facts. So, someone needs to work on this asap. Please discuss with @stelfrag to figure out who can help. 

@shyamvalsan there is another interesting finding in this discussion. The exclusion list for metrics is wrong. It excludes all the applications by default and includes many metrics that are just constant over time. We need to find a solution for this. I would suggest to open another ticket to discuss it and come with reasonable solutions.

**Reply by shyamvalsan** [reply]
Opened a new ticket #12173  specifically to measure the impact.

**Reply by shyamvalsan** [reply]
Also created a couple of tickets to cover the automating selection of metrics to train anomaly detection for. 

#12177 Ensure that "charts to skip from training" excludes only charts that anomaly detection will not add value 
#12178 [POC] Identify an optimal method to auto select what dimensions to train anomaly detection on 

And once the POC results are available we can create a separate ticket to actually make the change on the agent side.

cc: @ktsaou @cpipilas
